### PR TITLE
Add op bitwise_xor to replace __xor__ and __ixor__

### LINKED ISSUE #24513: Migrate `__ixor__` from the TH to Aten (CUDA)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

### LINKED ISSUE #24517: Migrate `__xor__` from the TH to Aten (CUDA)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

### LINKED ISSUE #24660: Migrate `__ixor__` from the TH to Aten (CPU)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

### LINKED ISSUE #24664: Migrate `__xor__` from the TH to Aten (CPU)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

### PR DESCRIPTION
**Pull request by xuhdev**
We define `bitwise_xor` instead of
`__xor__` and `__ixor__`. The reason is that (a) it is not idiomatic to call
functions starting and ending with double underscores, and that (b) the
types of argument that we can add is limited (e.g., no out), and that (c) consistent with the naming of `bitwise_not` and numpy.

Fix #24513,  Fix #24517, Fix #24660, Fix #24664

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
Hi! Can you please rebase.

**Comment by xuhdev** [comment]
@VitalyFedyunin I updated the PR to include a vectorized version on CPU.

**Comment by VitalyFedyunin** [comment]
Special thanks for taking care of the doc!

**Comment by VitalyFedyunin** [comment]
Code looks good, but need rebase to verify tests.

**Comment by xuhdev** [comment]
@VitalyFedyunin Done!

**Comment by VitalyFedyunin** [comment]
```
Sep 25 17:58:19 Traceback (most recent call last):
Sep 25 17:58:19   File "/var/lib/jenkins/workspace/cmake/../aten/src/ATen/gen.py", line 472, in <module>
Sep 25 17:58:19     generate_outputs()
Sep 25 17:58:19   File "/var/lib/jenkins/workspace/cmake/../aten/src/ATen/gen.py", line 464, in generate_outputs
Sep 25 17:58:19     "set environment variable GEN_TO_SOURCE or run \"{}\"".format(mismatch, update_cmd))
Sep 25 17:58:19 RuntimeError: Source files: ['OpsAlreadyMovedToC10.cpp'] did not match generated files.  To update the source files, set environment variable GEN_TO_SOURCE or run "cp /var/lib/jenkins/workspace/build/aten/src/ATen/core_tmp/OpsAlreadyMovedToC10.cpp /var/lib/jenkins/workspace/cmake/../aten/src/ATen/core"
```

**Comment by VitalyFedyunin** [comment]
looks like actual errors

**Comment by xuhdev** [comment]
@VitalyFedyunin Looks like the build issue is gone now

**Comment by xuhdev** [comment]
@VitalyFedyunin Rebased to resolve conflict

**Review by VitalyFedyunin — COMMENTED** [review]
Code looks good as soon as we win landing race ( aka need to be rebased again ).

**Comment by xuhdev** [comment]
Rebased

**Comment by xuhdev** [comment]
@VitalyFedyunin I rebased again, because there seemed to be a commit in master removed from commit history

**Comment by xuhdev** [comment]
@VitalyFedyunin Rebased again to resolve a conflict

**Review by VitalyFedyunin — APPROVED** [review]

File: aten/src/ATen/native/BinaryOps.cpp
Comment (VitalyFedyunin) [inline aten/src/ATen/native/BinaryOps.cpp]: Breaks XLA, please use `at::native::`
@ailzhang can we have use `at::native::` manifesto in wiki? ;)

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
xla fix required.

File: aten/src/ATen/native/BinaryOps.cpp
Comment (xuhdev) [reply aten/src/ATen/native/BinaryOps.cpp]: I believe this should be `at::` instead of `at::native::` :)

I've updated the PR

File: aten/src/ATen/native/native_functions.yaml
Comment (ailzhang) [inline aten/src/ATen/native/native_functions.yaml]: Hmmm if they're dispatching to the same function for different devices, then the dispatch is not really necessary?

File: aten/src/ATen/native/native_functions.yaml
Comment (ailzhang) [inline aten/src/ATen/native/native_functions.yaml]: same here.

File: aten/src/ATen/native/BinaryOps.cpp
Comment (ailzhang) [inline aten/src/ATen/native/BinaryOps.cpp]: I'm curious the perf impact of dispatching `scalar` to tensor version of functions. 
btw the failed XLA is not only XLA device, if you add a cpp test for `xor` with `Scalar`, it'll likely fail as well. (I tried commenting out the XLA part of test and it failed). 
```
    TEST_F(AtenXlaTensorTest, TestBitwiseXorScalar) {
      torch::Tensor lhs = torch::randint(0, std::numeric_limits<int32_t>::max(),
                                         {4, 2}, torch::TensorOptions(torch::kInt));
      torch::Scalar rhs(123456789);
      torch::Tensor result = lhs.__xor__(rhs);
```

**Review by ailzhang — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/native_functions.yaml
Comment (xuhdev) [reply aten/src/ATen/native/native_functions.yaml]: I think this will give an error message on device not supported, such as quantized. Am I right on this?

Comment (ailzhang) [reply aten/src/ATen/native/native_functions.yaml]: Oh! Hmmm I think creating device dispatch is an overkill to get error message, @VitalyFedyunin do you know how do we usually handle this ?

File: aten/src/ATen/native/BinaryOps.cpp
Comment (xuhdev) [reply aten/src/ATen/native/BinaryOps.cpp]: I have pushed an update -- this is indeed broken after replace `native::` with `at::`. The cause seems to be that the tensor returned by `wrapped_scalar_tensor` cannot survive some kinds of `at::` functions.

**Comment by xuhdev** [comment]
Note that currently CI is failing legitimately. I figured that this is caused by the improper definition of wrapped_scalar_tensor, which ignores device. I've made a PR (#28534) to fix this (this PR is now blocked by #28534).

**Review by VitalyFedyunin — DISMISSED** [review]

**Comment by xuhdev** [comment]
Seems #28534 is not the ultimate cause. I've changed `gpu_kernel` to `gpu_kernel_with_scalar`.

**Comment by xuhdev** [comment]
@VitalyFedyunin All checks have passed now

**Comment by xuhdev** [comment]
@ailzhang @VitalyFedyunin  All checks have passed now. Could you rereview? Thanks!

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by VitalyFedyunin** [comment]
Please rebase

**Comment by xuhdev** [comment]
Done

**Pull request closed without merging**
